### PR TITLE
Framework: upgrade from babel 6 to 7

### PR DESCRIPTION
**Pull request by samouri**
**summary**

Babel 7 has undergone ~40 iterations of alpha releases and another 40 of beta releases.  It is very stable and has numerous features that Calypso development can benefit from.  Those features include:

1.  Increased adherence to specification.  Notably this includes banning commas after a rest operator (not spread).
2. javascript configuration file (`.babelrc.js` vs. `.bablrc.json`)
2. Supports for the new React jsx shorthand for fragments: `<></>`
3. TypeScript support

Babel docs on the upgrade: http://new.babeljs.io/docs/en/next/v7-migration.htm

**changes made**
- Switched to the new naming scheme for packages: `babel-core` --> `@babel/core`, `babel-preset-env` --> `@babel/env`, `babel-preset-react` --> `@babel/react`, `babel-types` --> `@babel/types`, 
- Updated `babel-core` to the [bridge](https://github.com/babel/babel-bridge) which is necessary for anything having babel as a peer dependency -- aka jest.
- Manually specified the `transform-class-propertites` to be "loose" which mimics babel6 behavior and is much more concise than the new spec client format based on `Object.defineProperty` [1]
- fixed a single case of comma after rest operator
- switched from `babelrc.json` --> `babelrc.js` which simplifies the webpack configs.

Question for the reviewer:
 - is there any reason we had been explicitly removing the babel-lodash-es plugin from `env.test` within webpack considering it should be ignored by babel unless `NODE_ENV=test`

**to test**

- [x] docker
- [ ] desktop
- [ ] icfy build size

---

**appendix**
[1]: Here is an illustration of the loose and strict methods for transforming class variables.

For the following es6 code
```js
class Chicken { animal = true }
```

Loose babel transformation would get us: (default in babel 6)
```js
var Chicken = function() {
 this.animal = true;
}
```

Strict transformation would return more verbose code:
```js
var Chicken = function() {
  Object.defineProperty( this, "animal", {
    value: true,
    writable: true,
    configurable: false, // has some implications about deleting from object, modifying getters setters, etc.
    enumerable: true,
  } );
}
```

### CODE REVIEW

**Comment by samouri** [comment]
having lots of trouble getting the `jest` runner to transform files.

**Comment by samouri** [comment]
async-load chunks are currently broken -- will investigate

File: server/bundler/babel/babel-loader-cache-identifier/index.js
Comment (samouri) [inline server/bundler/babel/babel-loader-cache-identifier/index.js]: can make this diff just to `.babelrc` --> `.babel.js`

File: server/bundler/babel/babel-lodash-es/test/index.js
Comment (samouri) [inline server/bundler/babel/babel-lodash-es/test/index.js]: this was strange and I'd like to investigate more. For some reason babel was using double quotes instead of single quotes which causes funny escaping  + breaks the test

**Comment by samouri** [comment]
**build times**

I ran each of these measurements only once on my 2016 Macbook Pro and recorded the results. The number was taken from webpacks "Time" output in the terminal.
No deal breakers here -- babel 7 appears to be significantly faster in the server uncached and client cached cases.  I found it eerie how close the two measurements for client uncached were. 

| Test    | server uncached | server cached | client uncached | client cached | incremental |
| ------- | --------------- | ------------- | --------------- | ------------- | ----------- |
| babel 6 | 21790ms         | 11796ms       | 64486ms         | 55742ms       | ~5s         |
| babel 7 | 16623ms         | 12023ms       | 64872ms         | 44904ms       | ~5s         |

Note: incremental builds vary wildly even when just modifying the same file over and over.



**bundle sizes**

Stat sizes increased since there are more comments in babel7 output.  The actual gzip size became a bit smaller across the board :tada:

ICFY: http://iscalypsofastyet.com/branch?branch=upgrade/babel

```
Commit: 86d45c202aedff848b53b38db7c3a9ef21e4e106
Author: samouri
At: 2018-03-14T21:16:56Z
Message: the looser the better
Delta:
chunk                                                       stat_size            parsed_size            gzip_size
accept-invite                                                +10623 B   (+7.7%)      +1674 B   (+2.0%)     -235 B   (-1.3%)
account-recovery                                              +1297 B   (+2.2%)      +1122 B   (+4.2%)      -88 B   (-1.3%)
ads                                                            +956 B   (+0.8%)       -140 B   (-0.2%)     -153 B   (-1.0%)
async-load-blocks-calendar-popover                            +4533 B   (+1.3%)      +1246 B   (+0.7%)     -136 B   (-0.4%)
async-load-blocks-reader-full-post                            +6042 B   (+7.1%)      +2489 B   (+5.5%)       +5 B   (+0.0%)
async-load-components-happychat                                -729 B   (-1.1%)      -1488 B   (-5.1%)     -365 B   (-4.3%)
async-load-components-web-preview-component                   +8966 B   (+2.1%)        -19 B   (-0.0%)     -567 B   (-1.1%)
async-load-components-webpack-build-monitor                     -50 B   (-1.1%)       -191 B   (-6.8%)      -63 B   (-5.0%)
async-load-design                                            +38758 B   (+2.2%)     +13773 B   (+1.1%)    -1007 B   (-0.4%)
async-load-design-blocks                                     +45330 B   (+1.8%)      +9345 B   (+0.6%)    -1272 B   (-0.4%)
async-load-design-typography                                   -700 B  (-14.9%)       -241 B   (-7.6%)      -75 B   (-6.4%)
async-load-docs-selectors                                      -105 B   (-0.9%)       +229 B   (+3.9%)      -30 B   (-1.7%)
async-load-extensions-woocommerce-app-store-stats             +2807 B   (+0.9%)      +1230 B   (+1.0%)      -65 B   (-0.2%)
async-load-extensions-woocommerce-app-store-stats-listview     +239 B   (+0.9%)       -390 B   (-2.2%)      -37 B   (-0.9%)
async-load-featured-image                                       +95 B   (+1.7%)        -74 B   (-1.9%)      -32 B   (-2.4%)
async-load-layout-masterbar-drafts-popover                     -584 B   (-3.8%)       -432 B   (-4.5%)      -62 B   (-2.0%)
async-load-lib-abtest-test-helper                              -395 B   (-8.9%)        -96 B   (-3.7%)      -63 B   (-6.0%)
async-load-lib-happychat-connection                           +3147 B   (+1.8%)      +1154 B   (+1.7%)     -242 B   (-1.2%)
async-load-lib-network-connection                              +218 B   (+5.8%)         +0 B                 +0 B
async-load-lib-preferences-helper                             -2434 B  (-18.3%)       -247 B   (-4.6%)      -79 B   (-6.3%)
async-load-lib-rubberband-scroll-disable                         +7 B   (+1.5%)         +0 B                 +0 B
async-load-my-sites-current-site-domain-warnings              +5454 B  (+11.0%)       +902 B   (+3.0%)      -64 B   (-0.9%)
async-load-my-sites-guided-transfer                           -1139 B   (-2.6%)       +111 B   (+0.5%)     -139 B   (-2.6%)
async-load-my-sites-site-settings-section-export              -2903 B   (-5.0%)      -1240 B   (-4.3%)     -173 B   (-2.4%)
async-load-my-sites-site-settings-section-import             +10138 B   (+7.0%)      +3346 B   (+4.3%)     -205 B   (-1.1%)
async-load-notifications                                      -5213 B   (-2.1%)      +2334 B   (+2.1%)      -51 B   (-0.2%)
async-load-post-editor-editor-author                          +1715 B   (+5.5%)         +9 B   (+0.0%)     -133 B   (-2.7%)
async-load-post-editor-editor-discussion                       +233 B   (+3.5%)        +22 B   (+0.5%)      -41 B   (-2.8%)
async-load-post-editor-editor-location                        +1703 B  (+13.4%)       +890 B  (+12.2%)      -27 B   (-1.2%)
async-load-post-editor-editor-post-formats-accordion           +685 B   (+5.3%)       +163 B   (+2.4%)      -15 B   (-0.7%)
async-load-post-editor-editor-seo-accordion                     -11 B   (-0.2%)       -171 B   (-3.2%)      -80 B   (-4.5%)
async-load-post-editor-editor-sharing-accordion               +3757 B  (+10.5%)      +1622 B   (+8.0%)      -33 B   (-0.7%)
async-load-post-editor-media-modal                           +31398 B   (+5.5%)     +12117 B   (+4.2%)      +46 B   (+0.1%)
async-load-reader-conversations-stream                         -142 B   (-1.6%)         +4 B   (+0.1%)      -63 B   (-3.5%)
async-load-reader-feed-stream                                 +4143 B   (+7.3%)      +1484 B   (+4.2%)       -9 B   (-0.1%)
async-load-reader-following-manage                            +7350 B   (+1.6%)      +3260 B   (+1.6%)      +18 B   (+0.0%)
async-load-reader-list-stream                                    +5 B   (+0.0%)        +27 B   (+0.2%)      -49 B   (-1.6%)
async-load-reader-search-stream                               +5687 B   (+1.3%)      +2584 B   (+1.4%)      +51 B   (+0.1%)
async-load-reader-sidebar                                      -292 B   (-0.5%)      +1011 B   (+3.5%)      -57 B   (-0.8%)
async-load-reader-site-stream                                 +4567 B   (+7.5%)      +1547 B   (+4.1%)      -41 B   (-0.5%)
async-load-reader-tag-stream-main                              -233 B   (-0.8%)       +177 B   (+1.2%)      -82 B   (-2.0%)
async-load-reader-team-main                                      +9 B   (+2.5%)         +1 B   (+0.3%)       +2 B   (+0.7%)
auth                                                             +6 B   (+1.0%)         +0 B                 +0 B
build                                                        +58724 B   (+1.6%)     -24663 B   (-1.6%)    -1350 B   (-0.4%)
checklist                                                      +407 B   (+0.9%)       -361 B   (-1.2%)      -99 B   (-1.4%)
checkout                                                     +28661 B   (+3.0%)      -1345 B   (-0.3%)     -784 B   (-0.7%)
comments                                                     +22423 B   (+6.2%)      +8972 B   (+5.1%)     +257 B   (+0.6%)
concierge                                                     -4733 B   (-5.4%)        +89 B   (+0.2%)     -176 B   (-2.0%)
customize                                                     +4066 B   (+4.2%)        +94 B   (+0.2%)      -59 B   (-0.6%)
devdocs                                                       +6816 B   (+1.8%)       +587 B   (+0.3%)     -498 B   (-1.2%)
domain-connect-authorize                                      +1549 B   (+7.2%)      +1283 B  (+11.5%)       -3 B   (-0.1%)
domains                                                      +46522 B   (+4.0%)     +13432 B   (+2.1%)     -433 B   (-0.3%)
google-my-business                                              -93 B   (-0.7%)        +37 B   (+0.4%)      -78 B   (-2.9%)
happychat                                                    +11138 B   (+3.7%)      +2803 B   (+1.8%)     -390 B   (-1.1%)
hello-dolly                                                    -860 B   (-9.2%)       -507 B   (-7.7%)     -136 B   (-6.0%)
help                                                         +15693 B   (+4.0%)      +5704 B   (+2.7%)     -179 B   (-0.4%)
jetpack-connect                                               +8539 B   (+1.8%)       -719 B   (-0.3%)     -557 B   (-0.9%)
jetpack-onboarding                                             -354 B   (-0.4%)       +629 B   (+1.5%)     -133 B   (-1.3%)
login                                                         +3140 B   (+1.9%)      +2338 B   (+2.8%)     -167 B   (-0.9%)
mailing-lists                                                 +1662 B  (+17.1%)       +588 B   (+9.7%)      -42 B   (-2.2%)
manifest                                                         +0 B                   +0 B                 -1 B   (-0.0%)
me                                                           +12073 B   (+5.1%)      +4441 B   (+3.5%)      -64 B   (-0.2%)
media                                                        +24038 B   (+5.1%)      +8182 B   (+3.4%)      +44 B   (+0.1%)
notification-settings                                        +13387 B   (+3.2%)      +6373 B   (+2.8%)     -202 B   (-0.4%)
paladin                                                        -318 B  (-11.8%)       -283 B  (-16.6%)      -87 B  (-11.5%)
people                                                       +27860 B   (+7.8%)      +8766 B   (+4.6%)      -66 B   (-0.2%)
plans                                                         +9927 B   (+2.2%)      -3468 B   (-1.4%)     -506 B   (-0.9%)
plugins                                                      +30884 B   (+5.4%)      +3813 B   (+1.2%)     -118 B   (-0.2%)
post-editor                                                 +105828 B   (+4.3%)     +37086 B   (+3.4%)     +298 B   (+0.1%)
posts-custom                                                  +7893 B   (+2.2%)       -501 B   (-0.3%)     -675 B   (-1.7%)
posts-pages                                                  +11393 B   (+2.4%)       +667 B   (+0.3%)     -819 B   (-1.6%)
preview                                                        +651 B  (+10.1%)       +139 B   (+3.3%)      -49 B   (-2.9%)
privacy                                                      +12514 B   (+5.2%)      +4658 B   (+3.5%)      -42 B   (-0.1%)
purchases                                                    +43891 B   (+4.5%)      +8245 B   (+1.5%)     -165 B   (-0.1%)
reader                                                       +12585 B   (+2.2%)      -1764 B   (-0.6%)    -1027 B   (-1.4%)
security                                                     +30840 B   (+5.3%)     +15252 B   (+4.6%)     +258 B   (+0.4%)
sensei                                                         -349 B  (-10.7%)       -240 B   (-9.3%)      -91 B   (-9.8%)
settings                                                     +17114 B   (+3.2%)      +1404 B   (+0.5%)     -354 B   (-0.6%)
settings-discussion                                           +2932 B   (+2.0%)      -1671 B   (-1.9%)     -102 B   (-0.6%)
settings-security                                             +4092 B   (+2.1%)       -792 B   (-0.7%)     -194 B   (-0.9%)
settings-traffic                                              +5517 B   (+0.7%)      -1041 B   (-0.3%)     -271 B   (-0.4%)
settings-writing                                             +17496 B   (+3.3%)      +5098 B   (+1.9%)      -95 B   (-0.2%)
sharing                                                      +16402 B   (+5.8%)      +5604 B   (+3.8%)       +1 B   (+0.0%)
signup                                                       +34074 B   (+4.0%)      +8295 B   (+1.7%)     -489 B   (-0.4%)
stats                                                        +12591 B   (+1.8%)      +6968 B   (+2.1%)     -115 B   (-0.2%)
theme                                                        +14237 B   (+6.2%)      +2560 B   (+2.2%)       -3 B   (-0.0%)
themes                                                       +18508 B   (+5.2%)      +5166 B   (+3.0%)      -28 B   (-0.1%)
vendor                                                         +286 B   (+0.0%)         +2 B   (+0.0%)       +3 B   (+0.0%)
woocommerce                                                 +121971 B   (+4.6%)     +34095 B   (+3.0%)     +176 B   (+0.1%)
wp-job-manager                                                 -869 B   (-0.3%)       -332 B   (-0.2%)     -155 B   (-0.5%)
wp-super-cache                                                  +26 B   (+0.0%)      +1546 B   (+1.2%)     -248 B   (-0.9%)
zoninator                                                     +9421 B   (+3.2%)      +2724 B   (+2.1%)      -56 B   (-0.2%)
account                                                      +14861 B   (+3.9%)      +4488 B   (+2.2%)       +5 B   (+0.0%)
```

**Comment by samouri** [comment]
One more update:

I've also made the `template-literal-transform` 'loose' instead of 'spec'.  The behavior in babel 6 was somewhere in between the two.

original: `\`hello ${firstName}\``;
loose:  `'hello'+name`;
spec: `'hello'.concat(name)`;

**Comment by jsnajdr** [comment]
> First I removed the regenerator-runtime which we aren't using (generator/async polyfill).

Does that mean that async/await will no longer work in Calypso? That's bad news, I've been using it in the WIP version of #23090 😢

> is there any reason we had been explicitly removing the babel-lodash-es plugin from env.test

`NODE_ENV=test` is set in two cases: when running tests in Jest, and when building the server bundle in test environment (wpdesktop does that). See https://github.com/Automattic/wp-calypso/pull/22187#issuecomment-364036709 and https://github.com/Automattic/wp-calypso/pull/22187/commits/081bf996724b1ef040d696ae38270b7142c63abc

File: client/lib/users/test/fixtures/polling-users.js
Comment (jsnajdr) [inline client/lib/users/test/fixtures/polling-users.js]: What's wrong with `Array.concat` here? Maybe it should have been `Array.prototype.concat`? Or `lodash.concat`?

Comment (blowery) [reply client/lib/users/test/fixtures/polling-users.js]: `Array.concat` doesn't exist, does it? `concat` is only available on `Array.prototype`.

Comment (samouri) [reply client/lib/users/test/fixtures/polling-users.js]: I'm not certain why it was working with babel 6.  `Array.concat` is not a function :)

**Comment by samouri** [comment]
> NODE_ENV=test is set in two cases: when running tests in Jest, and when building the server bundle in test environment (wpdesktop does that). See #22187 (comment) and 081bf99

ah! so it turns out we can separately set `BABEL_ENV` from `NODE_ENV`.  I can add a `BABEL_CONFIG` flag to the desktop commands

**Comment by blowery** [comment]
Should we also look at taking advantage of babel-polyfill now? Or wait till after 7 lands?

**Comment by samouri** [comment]
> Should we also look at taking advantage of babel-polyfill now? Or wait till after 7 lands?

Definitely after! There are so many knobs to tweak and things to explore but theres no reason to do all of it in this PR

**Comment by samouri** [comment]
Just tested out the desktop branch and it wfm: https://github.com/Automattic/wp-desktop/pull/423. I'll merge that after this.

cc @astralbodies

**Review by jsnajdr — APPROVED** [review]
Works well in all possible scenarios. Reshrinkwrap and :ship:!

File: .babelrc.js
Comment (ockham) [inline .babelrc.js]: Tho https://nodejs.org/api/esm.html 😁

Comment (samouri) [reply .babelrc.js]: thats exciting! do you know when it will become stable?

**Pull request merged**
